aten/src/ATen/native/SparseMM.cpp

```diff
@@ -0,0 +1,83 @@
+#include "ATen/ATen.h"
+#include "ATen/Dispatch.h"
+#include "ATen/NativeFunctions.h"
+
+#include <TH/TH.h>
+#undef THNN_
+#include <THS/THS.h>
+
+#include "ATen/CPUFloatTensor.h"
+#include "ATen/CPUDoubleTensor.h"
+#include "ATen/SparseCPUFloatTensor.h"
+#include "ATen/SparseCPUDoubleTensor.h"
+
+namespace at {
+namespace native {
+
+// Calling into TH for sspaddmm because ATen code generation currently
+// doesn't support Sparse x Dense operations on Sparse tensors
+template <class scalar_t>
+void sspaddmm_TH_dispatch(Tensor & result, Scalar beta, const Tensor& self,
+    Scalar alpha, const Tensor& mat1, const Tensor& mat2) {
+  runtime_error("sspaddmm NYI for types %s %s %s",
+      self.type().toString(), mat1.type().toString(), mat2.type().toString());
+}
+
+template <>
+void sspaddmm_TH_dispatch<float>(Tensor& result, Scalar beta, const Tensor& self,
+    Scalar alpha, const Tensor& mat1, const Tensor& mat2) {
+  auto result_ = checked_cast_tensor<SparseCPUFloatTensor>(result.pImpl,"result",0, false);
```

Comment:
nit: it might not be that hard to avoid duplicating this code; just dispatch to a templatized function that e.g. for float does <float, SparseCPUFloatTensor, CPUFloatTensor>.  Can you can use <float>.to() instead of toFloat().